design-proposals/shadow-node.md

```diff
@@ -0,0 +1,343 @@
+  * [labels](https://github.com/kubevirt/kubevirt/blob/5b61a18f4c3b6c549d50fd201664c21720e3dbe2/pkg/virt-handler/heartbeat/heartbeat.go#L96) the node with `kubevirt.io/schedulable=false` when the stop channel closes (i.e. virt-handler exits).
+  * [labels and annotates](https://github.com/kubevirt/kubevirt/blob/5b61a18f4c3b6c549d50fd201664c21720e3dbe2/pkg/virt-handler/heartbeat/heartbeat.go#L139) the node, updating NodeSchedulable, CPUManager, KSMEnabledLabel labels and VirtHandlerHeartbeat KSMHandlerManagedAnnotation annotations once per minute.
+    * On the other side - Virt-controller's node-controller watches which monitors the heartbeat annotation.
+
+3. Virt-handler/node labeller [labels](https://github.com/kubevirt/kubevirt/blob/5b61a18f4c3b6c549d50fd201664c21720e3dbe2/pkg/virt-handler/node-labeller/node_labeller.go#L254) various information which are mostly static in nature: cpu model, hyperv features, cpu timer, sev/-es, RT...
+  * This is triggered both by interval (3 min) and on changes of cluster config. However, as these labels usually don't change—this intervalled reconciliation is usually no-op.
+
+## Solution option #1: Use Validating Admission Policy
+Add new [validationAdmissionPolicy and ValidatingAdmissionPolicyBinding](https://kubernetes.io/docs/reference/access-authn-authz/validating-admission-policy/) objects to prevent virt-handler from adding/deleting/updating non kubevirt-owned labels/annotations.
+kubevirt-owned labels/annotations are defined as follows:
+- labels/annotations that contain the string `kubevirt.io`.
+- `cpu-manager` label, which is an exception to the above rule.
+
+The validationAdmissionPolicy and ValidatingAdmissionPolicyBinding objects will be configured with the following parameters:
+- [matchConstraints](https://kubernetes.io/docs/reference/access-authn-authz/validating-admission-policy/#getting-started-with-validating-admission-policy): apply only on Nodes; Operations: UPDATE
+- predefined [variables](https://kubernetes.io/docs/reference/access-authn-authz/validating-admission-policy/#variable-composition):
+  - `isPartitionedServiceAccount`: true if virt-handler is the user requesting the update.
+  - `oldLabels`: the nodes' labels before the update was issued
+  - `oldNonKubevirtLabels`: a sublist of the old labels that are not kubevirt owned
+  - `newLabels`: the nodes' labels after the update was issued
+  - `newNonKubevirtLabels`: a sublist of the new labels that are not kubevirt owned
+  - (same set of parameters for annotations)
+- [matchConditions](https://kubernetes.io/docs/reference/access-authn-authz/validating-admission-policy/#matching-requests-matchconditions) rules (only if all are "true" then the validation will be evaluated):
+  - `isPartitionedServiceAccount` == true
+- [validation](https://kubernetes.io/docs/reference/access-authn-authz/validating-admission-policy/#creating-a-validatingadmissionpolicy) rules (written in pseudocode to keep things readable):
```

Comment:
Is it possible to express something like: "reject-all-changes except these"?
I mean, without the need to compare other labels/annotation we do not care about.

Reply:
well, it was nice if we only got the "changes" we made, but it doesn't work like that AFAIK: When setting a POST request (this is true for both webhooks and validationAdmissionPolicies) you can add validation rules to validate [these](https://kubernetes.io/docs/reference/access-authn-authz/validating-admission-policy/#validation-expression) entities (such as `object`, `oldObject`, `request`  - see more details [here](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#webhook-request-and-response) to learn how they look like)
so it's logical that you first have to gather the kubevirt-labels on both {`object`,`oldObject`} (or the non-kubevirt-own, as shown) , and then make sure that we "reject-all-changes except these"..
Please review this CEL example that models the validations that the valiationAdmissionPolicy will have [here](https://playcel.undistro.io/?content=H4sIAAAAAAAAA%2B1da3PjNrL9K4w%2BbDJVJkcv27Kq%2FMGvycxev9ZyHO%2But1QQCUqM%2BBqAlO2k8t9vA6Qs6kFKaCpbt25hqnZjQjwHQANoNkg2zx8NhySk0W9Eo9%2BonfRfQgP%2Bkdh7oox7Udg3Zq2sbOqFTt%2B4jRyaHQc0IQKbQwAUhlFCEgDxjzLDsElMbC95t5IojvxZYHnR52bToS5J%2FaRvvDQ6vXbvsNU77vZaR82XRjVyNm5VY7hnzSI%2FDag1TUeUhTShXABDaLcH7f%2Fxj5fGgu%2Bl0X9p2OHM%2FE5NL3QZMdsdC46%2F07blk5FFw7HFnLRtMepMSGLZUfDS%2BPPHRX3THreiWWhFbPx5EvHEtD2Hcajm3y%2BNVtNqWa3DI6vd%2BdzuvjT%2BU4LzO%2BaYJPSVvJt2FLreOGtlbiJoIhwEwuyisXxCoC0vjYOXhhcmlLnEhqY78rcRM%2BnbULU7wBQQ2ySOwyjnkoh2%2B91OvzvqU7t%2FfNwXJhb1xfOTqDhtUw%2BXzpJUq6fAGSF9S8xJFK%2BRHOYU8xNWCA5zOJjCjCOWmDQkIz8zS8JSKn%2Bd%2BSSc2wOa%2FWfZYMHEeI3YFE7lq%2BbOgGU4Ubs9IZx73BQTqtMdHbbcbs9sdZots2v3Oubo%2BIgKI5BDu0Pb9lGngmo%2B8ixKYTRNn8Wm5woDZq3y4lk3N0PTOupaTav9uXW0pXmiWS%2BNdnFdrJ0TjIMks2Jh8PtGr91vt%2FtOu3886p%2B0%2Bj2ngiJmXkDYe1l7l8e9ssE8HcF4rA1EPjnaPeh283O7A7OjkiZhJOReYvJXL7EnWe9KjdnrbTUmo0GUUPP3KAR7eWMGQwWWVVxfG5klZUgCWoMO3NvMY4lwbhNKWDKiRPrTdrPdNZuHZrP12Gr1m8f99sm%2FlqZCATjlgTkhoePD1AtISMZUTh2X%2BJwWMTBHJl5IM%2FeUMunkrSimIZ94rmSCnxIW%2Bfew%2FOij8K%2FR%2BL1vDLxw7NMHGvueTZToUsZomFzk%2FhD%2BdihYARrJxTKx3eMOIXTUbp50em7z8OTk8MjuUjKizRNK6aFKVQ7lHlD%2FF6u6ZMSDC2sa2hFzotD8K6v0gfIsBvv%2F96u9G3HKZtQZiP9nF2dnH%2FEBco4J0sF7aMvBErMNpm02bA%2BURymz6UfQ8tLotjrdTqd3rFZFLK%2Fh8neT2PO2qnEwSjgGx8E44BAuwTcsYKHrWMK5rYQzLiz2lFET3AT1wXHacWrC%2FzzHOrt8PigcXQ1uvxWPn56Xj9rLh4et9vmva0UXl2tFl%2F9YK%2FqyVvJ0vV50u9Sei5v754uvP%2FcKRV%2Bufxl8HV63inV%2BuTnrFA%2BfBw8rh3f3j4WSb2ed9vDs4eLr8OLsvlh%2Bfn9eOLw%2B%2B1ps9M3l8OL66qxIfXP3dH5VPL4v2u9u8Dw4eyr%2BPri%2FuhhePD5cDweD82IPBo9Qd%2FH4n4OLs%2Bvr5ZKrItXTTbGq595x8ejnq8fzp1axZKUh8vhitWDZSrJokBXIyWdlEZUjiyDQdF4h2BwGcDX2kglMawfcufxNhKO%2B5ZLA898LBZ5TOJiBe4nYMC9jTmI9XD5e3DwWD687F2fF45vzlcObpcO724OpWAR%2BvkSt27vh139tKht%2B%2BQWMm%2F%2FAqe%2BF6dtH7%2FLiWeYsLDf1%2FdWygPwWsbVCL1wvZHAxFX8cBBAtQDQUpgE5mEeXnHnRzBIbGah4tTT3BNCg2PbMZqfZHLaa7ZEVQygG174DnkQMrshmGIUsd5zEdzw%2BPeDv4LADM%2BJDRn3wNtT6drmh8O7%2B6nbw9duXx%2BHT1cPgG1hvwzmDx4erq4oTHr5eXVf8nP8y3NiAxY%2B5QatPEebd6vmyi5U1m3v6FC5iLjhWZytS2H4LUo5MPkwbnfM3CJUWpycpoJfPG0VRYgcOTDg4m0%2Fp6zDx7Olpy0i4fQqd9sRMMJidxo5YcPDHEHodEH9IXOjXUOBPW4sqoDfR5PfTKDSyU%2B0RP%2B2anZODbtc8PslbADNk6PHIh7XCT5vin9tsHjSzf66RWd2x4GeoBboMm%2BrT5kHroH3QOeg2D7qtg277oNsp1io2l%2F7Qi4IgFZVnf8SJ8VFNHi4OPfb9oNAg0dqhWFGFVhZ5l8wQRjxyEz%2Byp2lshIE3fCUQuDvR%2BLRpBDY9jVx3yYZggoRRGNfMGw1hBxKdtpasle8ehpN0TGNoIP%2F9tPVz3ptYDuIpLCG5b%2FxAZTcM%2BMp0sT9ijHlobJIkgau46VDxH3GBz7adcyYbGiVW6aMHbAkJ4kU43jFbXRGOtzr9ZrMQjsNgEN%2F7Habkxy0T0yjc7wg%2FbrcYRn6tX7QbdjbU%2BZ88nL97hZlwLU6xChH%2BeiMNYwTNX%2BkrYaI%2FJHCOupXnRRBqSF9auOMC%2FjmPRhaLbr6%2F6Dhh9Ar%2BzKXJJoNtQZNRoI5x3tQxlKtjAsfkfOTggIk3itWRsWerg1i2K1QEwSyXVy3P9xIPY54ZYhhmb20MCKLK0SsSaCMGUAKd70igi8TNfCwwDD1l6CjwWhiQ%2BgDavuunXN055LgoVp%2Fetv86UgcF0Uwd9NZDYFpHyiCHKkMoC9TXtds6UvdB7si0Idpj6kAxwqbfUl%2BkbkDUMXGqjuFjPoI4Vh34xtXtMXY47AaUYZP3mDLYpUTqNcJ1Sn2peCOmPrMEyCS%2Bupfzwhnsn9SnCOAgGFWG%2BWTiDn31wAQBCWz1SQxhszrGQa7PADcfg0A9OADvO0L0LEbUhFiYQcLUQaF622KiboMYEQCiFlRs%2B0HqIwKj2LHVF0fsxLQ1VndO8Rhhwqn6lSFGRF1xFNshYrS4PSHhmJqBjVmOMXdiHA5hSU476tENc%2BzExzSROYyE6jOZOZxSDAwuKOr7OthFMthL4gaBU0SFUy8WUZXrm7OAhgl7V6cICKLeANPYmNqmnTD1wICrByGoHT1HrAPAqG%2BVANS11LdlEoapDdaqOgp1a4O%2FcxsT%2B2ECOMCYDiX5LVp18JA4v6Vc3Usn%2FA03j9PAU7foLFDv3AwRlwFGPiidecm7OfHVzbJE8ErE6xterH7plDSxZ89MRseeeERQg2KGiH4W6Lc26k5dgQCNt1nH9CPiwJWEvnm4wRAc4tETrUfSyxpSiyFrBprCobDU0Wh5YTQ90mlT%2BUSzBom0xCh0bHesfk1aoaEucmoXODBbglUKylxz7Ecj4uO82oIvjCSjQ0fpGOI8HBPiNmCOMzG7iDm2HeCx9E28DeKrB185AcE5KUDG%2BStyXjjGUcCaAqc9NT0YQDyDvOlQb10ImprL4oMCvSoWDHtZFIIujDiZ0ZprQhBJmnrGkRS1jJMxMEqDWDwfNRMvQDbI9embeOrLIFpAEUCIgr8keCE3ozTBzVWxVupUPcP6uAwq7q2KV2QS%2BlaLhstXOusy%2BfG4li3EDapaBLMYGePl2H2YU%2FLk9szeUsbwRObISzDb8RyNtmIQhR4EaHUIZk4dOGd1uh4k6o9iJU5uUtCtDgMPDxZvtMAmyZ%2BaXSQ%2B5TUC6jjAXYnEW7XUwUcLe7l0MCcOamwHsluJdfDinmIdfFJnN5Pdm8TjuYhVHZHuAeEI7gLIJ8SJXs1ZYCPxcjfoI57sCrQA4NsubB%2B5Lqc46wm4uMOFDbjTkFGeMM8Wy2icUsTtJ0EzkzFITEMH25BZ4KYh7t7ELHiFqI2KW84iJnQ96ju4wZiBD0UD63UfGzW8juB6X2P1v4kQWt1ayFtRb2OajGbqd5plK3EoRCMFDPP6T7Ut5cvrebqXeL97DX7OYKv3Sn3xQOpx8LwvHvPb%2BcMATXZBwL3AOWRKzSzfZo9UNTv6lfA9mGuJpZ6xvs3ez5nnlD8I3p2hXkNu6YT4tPTZ9674eo24ixPKonD4c%2Bl630pxT0P2HqLhAwis6o5IgaOeOQbTdzn1L3yPhsk%2BptucsbiY6jH%2BCgEAhMF4a80Jtjejnv%2Fbg9fbk4Oq45ZqOyOsC6rleHDupoaTQbsWlEPBu5GazmNfLmNfjgLpHpBOQe6%2FNz9Udxn9ntLQljdm282Tw%2BbxcbPZVGAQWySRmlNWeZZSuGFViPSeElCWu7MxHFqLEAvJP3EhnxaPFdm3ePRTjZqfnje%2B8bErWOYB1yS4uKxJcPmPmgRfauKfrusS3NYY%2F3luNJrgI5Maz3BztvEVoN3Az4OHWuC7%2B0c0fikDHM9yfn%2BOBovscjR4nouOJxCZ63j0Pd735FnxaPxyDj2eRmTc49FZfn4d%2FBXeBE83ePM%2F947x2Oy7Anh8rYHPvllQD17HZ2TfQ8DAl76egCAo%2BdYCgqn4ZQYBP1LGZnc4e4fKwI8PPvQNGZDtDF98GwLR4cKXJPDom%2FNa4Jsa4LtbRfCGr13UZ5Dfy8DRrHxdA0dS%2FBZH3zi0Wl2raXbbxxb1T6y33tGwmJi%2BO5386oRo0a6zedMXPwS%2B1cURzL8OIjiaO1IUPiSiaMyNnx2px%2FHxkRJFmo2fNFHkKP0AiirPhs%2Bl9A02sSOOxK99WUW0qGu1dvW2277D0je6rSPrpGuJr0eAGXutVrNjNpHcxU%2B4iIaeWLvO5qrvtdTu8urXYSThXtg%2BVu1S4%2BT3ObNbo%2BCAsc9QCizifof4ntv8URNf56z6TASCqSy5QpFql887qFFWpXgrMlXlfKtRladLq%2FFU50%2BrcZUlVKuybM6wVmcpS7lWZCrNpVbjqcpEVmXaz9iX5aYqsmxOIFUk2ZrmqcpXmnKoRlSVnqnGtC0VUpFtpzxHRc49rLs9ufHqXDRVrvLkNDWmsmw1NZaSVDRlkq25aTUYK5PVELxbs9ewnCXpbFi68peK0Iz7I9whAw5Hui0lDsO6JUcOR1mdNKfOWZlFp063S1odlnVbnl0t3rIkmlqkJWk19Th3yELCVrBDrh6CevNLfCiismw%2BHFlJeh%2BOrCrfD8W4OQEQRbUlIxDBuS1FEEm5NWcQybvvJb4tq7AW5V%2BzwHfMO0QyVyYi1uHcr4F3TlVU59%2BWu6jOWJXMqM5Wld2IYatId8TQleU%2FYrm2ZfChebdnSKKoq1ImUYSVOZQoxrL0CDTZXzJGu6RdIoir8jBRdPsbmm2ZmhjGqtRNBF9lLieCb3NyJ4KoMttTna8q%2FVOdbUs%2BKIawKkEUwbc5YxRBVJ1CiiD8S67x1UmmGL7KrFMMYWUaKoawKi8VyVeeqKpOuD1zFcFZncqKIKzKbVWnEwx77G519iuOryIdVp1wt%2FxYdd6tCbMIytIMWgzXjim1COqSHFsc055NuLcwc0tarjpheW6pGlcFj3wYvY4o%2Bwp52fnzbI2yx%2F8lsLLHHSWnw0U29L3xJKFhUPICUDm0zN%2BUAdJQXMlVIPw9LHnwUAXYmFlRiSiPMEpQqh1J%2FFGpjEIZBpZR6NCNz9mWX7bZqA6zfIqY3fX1LCuFZopNF7qVhfcN17QEl%2BcR8efmLDuH2xPqpKW5R8J0Jov81beQcokgMxaql6v8ZaBMwGrXszPRqk1nL%2BtORVy%2BeLvyUlveL%2FC8Gx8fL1K2zEW%2B1iYDJLmip7UiTYQdbvxUEZ8D7ht%2F%2FJkfJeBAF5JIxPcjmyRyGBeNj6Psotg%2BlG9gvoSR79xptWutdq3VrrXatVa71mrXWu1aq11rtWutdq3VrrXatVa71mrXWu1aq11rtev%2Fa2rXYn8y03rXqxitd70M0nrXVUCtd70ZqvWu10CB1rvWetdrGK13vQmk9a7XcVrv%2BgOm9a5ldK%2F1rnOg1rvOMVrveiNK612Xg7XedSlS610vYFrvuviel9a73ozRetda71rrXe%2F%2BYY4dSbTedRWf1rvGEWi9a613vZVI610vE2i9a613rfWui2itd62I03rXyjitd631rrXetda71nrXWu96M0rrXWu9a613jcNrvWutd40j0HrXWu96R5TWu9Z617HWu1bDar3rOgRa71rrXWPxWu9a611rvWtlrNa71nrXWu9aFaz1rrXetda7XqLRetda73pnKq13rfWuP5i03rXWu94Hk9a71nrXWu96Zy6td631rrXedTmp1rveD53Wu9Z616vUWu9a611rvevdKbXetda73pbMqPWutd51NaHWu9Z613hGrXeN4tN610girXe9J0Ktd70HTq13XZdP613vg0vrXddi0nrXWu9a612vQ5ZfttF615HWu%2F5%2FpHctO80ztWuWa0fm549ZlIotqDyYrYpgf5wsByCDS6oHPIscLaw9BI7HRHB9zAYTYOI3KGC53uYv95dnj1eiEHZ37FvoRnlDxWFWd%2FZCXV9odno2JbYdgTftz%2BfLxx9zvdgcLqZcs9V1R657QprHdlZM3xJWEAZPkwm4fs%2FOcpGXp7ec9rIFi%2FEylQ3xEjYOGuLmKuXC1I1%2B4%2FNn41fqw48U5q4BDTAurq6Ne5%2B8i7EJnR9eQjhluczwuEFCqRbG5EN%2BChyjM85pMPLfYRq9Cr1Ug4ZgiEhezgQ3VOtD7A5IR%2FxNmSd%2FefWSSVZvFARRaFx9tM64JuE4hY278RPU%2F8kSLZGtMY1fRTRnvMPskE1b9MiYQNX5Sb9wKokJODYDfhJ%2Fc8%2BhhhsxaHycJoYQZj%2BAv41%2Fnt1cG1D898Hdrfg9IEnOci%2BYjR8f0vBH2YsZ8VNSUjkZEy%2FkiaxpUUFOdJX3X%2BBsIZgmdVONyAUGEsQ%2B5XnDeOxl81GMF0BD%2BnobJWewZF9z4TJu9E%2BNSCrRW3N9eSvTPbNcz4dh%2BWl6YPwwFW8lJ6JJP700Ch7tpfFJEke%2Bs5F4rnKP5pbsjAoO6YeM7H0sWMF5o2XhabEqUbIClC5rAf3b3%2BZdzn9ZxsuyZYYcKsZ33hODp%2FbEIDB%2FHRFmF4TlDJrYlpVVs26ARc%2Bnxg%2Bn4CMzi7w0DDg%2FLyLhXLh3uTyXxfsiNx1Lv7BlPV6wnQWeWdSy3oR%2FT%2F8je7xW%2BmlTp0Wfwyg05yOTqeJBaCK8lewkh17%2FVGsOffokGpTx1Jwyn1Q6MXo3pvRdrFaxFmlxZiBrn1t9Kmqs2xejtDnzISyrQAwmGKLx5%2F8Cu0lKcVz9AAA%3D)
Please play with it and if you find a better way to do it I'm all ears of course :)